substrate/frame/broker/src/tests.rs

```diff
@@ -891,6 +891,164 @@ fn short_leases_are_cleaned() {
 	});
 }
 
+#[test]
+fn leases_can_be_renewed() {
+	TestExt::new().endow(1, 1000).execute_with(|| {
+		// Timeslice period is 2.
+		//
+		// Sale 1 starts at block 7, Sale 2 starts at 13.
+
+		// Add a core.
+		assert_ok!(Broker::request_core_count(RuntimeOrigin::root(), 1));
+
+		// Set lease to expire in sale 1 and start sales.
+		assert_ok!(Broker::do_set_lease(2001, 9));
+		assert_eq!(Leases::<Test>::get().len(), 1);
+		// Start the sales but don't offer any more cores.
+		assert_ok!(Broker::do_start_sales(100, 0));
+
+		// Advance to sale period 1, we should get an AllowedRenewal for task 2001 for the next
+		// sale.
+		advance_to(7);
+		assert_eq!(
+			AllowedRenewals::<Test>::get(AllowedRenewalId { core: 0, when: 10 }),
+			Some(AllowedRenewalRecord {
+				price: 100,
+				completion: CompletionStatus::Complete(
+					vec![ScheduleItem { mask: CoreMask::complete(), assignment: Task(2001) }]
+						.try_into()
+						.unwrap()
+				)
+			})
+		);
+		// And the lease has been removed from storage.
+		assert_eq!(Leases::<Test>::get().len(), 0);
+
+		// Advance to sale period 2, where we can renew.
+		advance_to(13);
+		assert_ok!(Broker::do_renew(1, 0));
```

Comment:
do_renew actually checks [cores for sale](https://github.com/paritytech/polkadot-sdk/blob/1c85bfe901741f9c456af1ac92008d647660a2f4/substrate/frame/broker/src/dispatchable_impls.rs#L134).

This only works, because of the request_core_count above, which takes effect after the `do_start_sales`. The 0 is basically overridden more or less immediately.